src/api/v1/db/oracledb/pizzas-dao.js

```diff
@@ -348,4 +365,151 @@ const postPizza = async (body) => {
   }
 };
 
-export { getPizzas, getPizzaById, postPizza };
+/**
+ * Perform a full replace of the ingredients relations of a pizza
+ * with `pizzaId` using `ingredientsQueries`
+ *
+ * A member of `ingredientsQueries` should look like:
+ *
+ * ```sql
+ *  INTO PIZZA_INGREDIENTS  (INGREDIENT_ID, PIZZA_ID)
+ *    VALUES (:ingredientIdX, :pizzaId)
+ * ```
+ *
+ *  Where `X` is the ID of the target ingredient resource and `ingredientIdX` also
+ *  appears in `ingredientsBindParams`
+ *
+ * @param {string} pizzaId
+ * @param {string[]} ingredientsQueries
+ * @param {object} ingredientsBindParams
+ * @param {OracleDB.connection} connection
+ */
+const replaceIngredients = async (
+  pizzaId, ingredientsQueries, ingredientsBindParams, connection,
+) => {
+  ingredientsBindParams.pizzaId = pizzaId;
+  const deletePizzaIngredientsQuery = dedent`
+    DELETE FROM PIZZA_INGREDIENTS WHERE PIZZA_ID = :pizzaId
+  `;
+  await connection.execute(deletePizzaIngredientsQuery, { pizzaId }, { autoCommit: true });
```

Comment:
I'm open to this change. My concerns are:
* in theory, we've already validated the ingredient IDs and can guarantee that they're safe
* I've had bad experiences with not committing changes, including a situation where what I think happened was that a connection made uncommitted changes, then crashed, and then held a lock on on the resources it changed for like ~24 hours and made it impossible to update them

Thoughts on that? Given the combo of 1 and 2 I think the current design is _relatively_ safe. But partly that's because I have very little confidence around working with oracledb commits.

Reply:
My bad, I didn't mean that there shouldn't be a commit, just that the commit should only occur in the last statement (the insertIngredients one). The main concern is that if the deleteIngredients statement succeeds, is committed, then the insertIngredients statement fails, I'm pretty sure this would end up just deleting the ingredients for a pizza when a patch was requested. We would want to respond with an error like a 500 here, but as a user, I would expect nothing to have changed.

So I think if we execute the first statement without committing, then only commit after the second statement, this will be prevented. Then, if an error occurred in the second statement after a successful first statement, the `finally` block would close the connection and [rollback any uncommitted changes](https://oracle.github.io/node-oracledb/doc/api.html#connectionclose).

Reply:
> Then, if an error occurred in the second statement after a successful first statement, the finally block would close the connection and rollback any uncommitted changes.

Cool, this basically answers my main concern. I'll push up the changes now.